### PR TITLE
[nuvo] Fixes protocol errors when connecting via an MPS4

### PR DESCRIPTION
**Pull request by boc-tothefuture**
Signed-off-by: Brian O'Connell <boc-tothefuture@users.noreply.github.com>

fix[nuvo] fixes protocol errors when connecting via an MPS4

This fixes the error when using the Nuvo binding over IP through an MPS4 the binding would be disconnected over and over again.  The MPS4 will disconnect connections when requests are sent and no zones are on. The MPS4 sends a PING command every 10 seconds, this update listens for that ping and if a ping is not received in minute initiates a reconnection if the user has indicated they are using the MPS4.

This issue is noted by another user in the forum [here](https://community.openhab.org/t/nuvo-wha-grand-concerto-mpsv2-binding-disconnecting/119410)

### CODE REVIEW

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java
Comment (mlobstein) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java:161]: Just a thought... Instead of having a configuration flag, why not just check if the port is 5006 and set isMps4 accordingly? It would make it easier for the user by auto-detecting the condition that requires the protocol change. The odds of the MPS4 port changing from 5006 in the future is non-existent.

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoConnector.java
Comment (mlobstein) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoConnector.java:308]: change to:
`if (PING.equals(message))`

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java
Comment (mlobstein) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java:596]: change to:
`if (!isMps4) {`

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java
Comment (mlobstein) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java:616]: change to:
`if (isMps4) {`

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java
Comment (mlobstein) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java:707]: change to:
`if (isMps4) {`

**Review by mlobstein — COMMENTED** [review]
looks good!

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoConnector.java
Comment (boc-tothefuture) [reply bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoConnector.java:308]: Sometimes the ping message is preceded by some random characters.  Which is why I did contains.

Comment (boc-tothefuture) [reply bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoConnector.java:308]: I could make it 'endsWith' which would cover that case and also not capture a "#PING" in some other message somehow..

Comment (mlobstein) [reply bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoConnector.java:308]: I had not seen that in the experiments I have done with trapping and responding to pings for NuvoNet source communication. Just keep it as contains() just in case.

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java
Comment (boc-tothefuture) [reply bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java:161]: Sounds good..  Given legrand shutdown the product line - the chance that anything changes in the future is pretty low :-)

**Comment by boc-tothefuture** [comment]
@mlobstein I believe I have resolved all your comments. 

Thanks for the review.

File: bundles/org.openhab.binding.nuvo/README.md
Comment (mlobstein) [inline bundles/org.openhab.binding.nuvo/README.md:51]: This is a little awkward,  how about this:

> * MPS4 connections do not support SxDISPINFO commands including those outlined in the advanced rules section below.

Comment (boc-tothefuture) [reply bundles/org.openhab.binding.nuvo/README.md]: done

**Comment by boc-tothefuture** [comment]
@mlobstein it may be worth you trying this build against your system?  I don't know if you are serial or IP over serial but I am not able to test that.

**Comment by mlobstein** [comment]
> 
> 
> @mlobstein it may be worth you trying this build against your system? I don't know if you are serial or IP over serial but I am not able to test that.

I will test it out later tonight and report back if there are any issues.

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java
Comment (mlobstein) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java:195]: I found a small issue... 'port' needs to be checked for null. When port is left blank in the configuration it throws an NPE here.
```suggestion
        this.isMps4 = (port != null && port.intValue() == MPS4_PORT);
```

**Review by mlobstein — CHANGES_REQUESTED** [review]
One small issue found. Otherwise serial and MPS4 connections are working as expected.

**Comment by boc-tothefuture** [comment]
Merged your change in.

**Comment by boc-tothefuture** [comment]
@mlobstein the PR still says that a review is needed.  Can you approve it if you are ok with it being merged?

**Review by mlobstein — APPROVED** [review]
@boc-tothefuture I can't merge the PR, only one of the maintainers can. I will mark the changes as approved FWIW.
@fwolter can you take a look when you have time?

**Review by fwolter — APPROVED** [review]

**Pull request merged**
